File: 백준/Gold/1806. 부분합/부분합.py
import sys
input = sys.stdin.readline

# n은 수열의 길이, s는 만들어야 하는 부분합
n, s = map(int, input().split())
# arr는 수열
arr = list(map(int, input().split()))

# length는 만들 수 있는 길이들을 담은 배열
length = []
# 한 개의 숫자로 s를 넘을 수 있는 경우
if max(arr) >= s:
    print(1)
else:
    first = 0
    second = 1
    # cur_sum은 부분합
    cur_sum = sum(arr[first:second+1])
    while second != n:
        if cur_sum >= s:
            length.append(second - first + 1)
            if (second - first + 1) == 2:
                break
            cur_sum -= arr[first]
            first += 1
        else:
            second += 1
            if second < n:
                cur_sum += arr[second]
    print(min(length) if length else 0)

File: 백준/Gold/1806. 부분합/부분합.java
import java.io.*;
import java.util.*;
public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int N = Integer.parseInt(st.nextToken());
        int S = Integer.parseInt(st.nextToken());

        //int[] arr = new int[N];
        int[] sum = new int[N + 1];
        st = new StringTokenizer(br.readLine());
        sum[0] = 0;
        for (int i = 1; i < N + 1; i++) {
            sum[i] = sum[i - 1] + Integer.parseInt(st.nextToken());
        }

        if (sum[N] < S) {
            System.out.println(0);
            System.exit(0);
        }

        int start = 0;
        int end = 1;
        int length = N;
        while (end <= N) {
            if (sum[end] - sum[start] >= S) {
                length = Math.min(length, end - start);
                start++;
            } else if (sum[end] - sum[start] < S) {
                end++;
            }
        }
        System.out.println(length);
    }
}
